FIPS/fip-0032.md

```diff
@@ -21,290 +21,426 @@ replaces: N/A
-up every pricing formula.
+Benchmarking overheads were measured and corrected for in the data.
 
 ## Backwards Compatibility
 
-The data model by which gas allowance and usage is represented in external
-interfaces (gas limit, gas used, etc.) remains unchanged. However, the same
-transaction made before and after this FIP will, with very high likelihood,
-result in different values for *gas used*. Transactions sitting in the mempool
-at the activation epoch of this FIP will need to be re-estimated and replaced.
+External-facing gas representation (gas limit, gas used, etc.) remains
+unchanged.
 
-## Test Cases
+However, the gas usage of most messages will increase upwards, so systems using
+static gas limits should be adapted accordingly.
+
+Clients sending messages near the activation epoch of this FIP should estimate
+gas under the new pricing model to guarantee inclusion.
 
-Testing for this FIP involves two efforts:
+- If the message is included pre-activation, the message will result in a higher
+  overestimation burn.
+- If the message is included post-activation, the message will be executed
+  normally.
 
-1. An updated corpus of FVM-compatible [test
-   vectors](https://github.com/filecoin-project/fvm-test-vectors) with the new
-   gas model.
-2. The testnet plan laid out in FIP-0031.
+## Test Cases
+
+Refer to test plan laid out in [FIP-0031].
 
 ## Security Considerations
 
-TODO draft.
+Metering and pricing execution enables more precise gas accounting, which
+translates into higher network security.
+
+Note that this FIP alone is insufficient to secure the network for
+user-programmed actors; it is merely a stepping stone. A future FIP will
+introduce further the final gas model changes before enabling user
+programmability.
 
 ## Incentive Considerations
 
-With the introduction of execution gas to account for real execution cost, actor
-compute logic will no longer be exempted from gas charges, thus resulting in a
-net increase in gas expenditure in the network.
+The gas utilization of all messages will increase, except for value transfers
+which are unaffected.
+
+In order to quantify the magnitude of change, we have backtested this model
+against 2.5MM mainnet messages.
+
+- Annex A includes a detailed breakdown of net utilization increase per message
+  type (relative to original gas).
+- Annex B includes the final total gas per message type.
+
+On average, most message types increase their utilization between 1.5-3x. The
+gas performance of specific messages types is subject to various factors. The
+reader is advised to study Annex A to understand the variability involved.
+
+In our backtest, the four most important messages for storage provider
+operations were impacted in this manner:
+
+- SubmitWindowPoSt (miner actor): gas usage increases an average of 3x, with a
+  median of 2.73x, and a 90% percentile score of 4.42x. Some executions spiked
+  as much as 14x.
```

Comment:
IMPORTANT: note that some window posts recover power, which has a significant computational cost that was not previously accounted for.

Reply:
Added.